File: 589.cpp
#include "lib.h"

namespace PE589 {
    const long L = 30, R = 60;

    double f[100];

    double F(int L, int R) {
        for (int T = 0; ; ++T) {
            double diff = 0;
            for (int x = 0; x <= R + 5; ++x) {
                double e = 5 * (R - L + 1) * (R - L + 1);
                for (int a = L; a <= R; ++a) {
                    for (int b = L; b <= R; ++b) {
                        if (a + 5 < x) {
                            e += a;
                        } else if (a <= x + b) {
                            e += a + f[x + b - a];
                        } else {
                            e += x + b + f[a - x - b];
                        }
                    }
                }
                double fx = e / (R - L + 1) / (R - L + 1);
                diff = max(diff, abs(f[x] - fx));
                f[x] = fx;
            }
            if (diff < 1e-6) {
                print("diff = {:.3f}, T = {}, ans = {}\n", diff, T, f[0] - 5);
                break;
            }
            // print("diff = {:.3f}, ans = {:.3f}\n", diff, f[0]);
        }
        return f[0] - 5;
    }

    void main() {
        double ans = 0;
        long k = 5;
        for (int R = 2; R <= k; ++R) {
            for (int L = 1; L < R; ++L) {
                ans += F(L, R);
            }
        }
        print("ans = {}\n", ans);
    }
}

int main() {
    PE589::main();
    return 0;
}

File: 589.py
import numpy as np
import numba as nb

@nb.jit(nb.float64(nb.int32, nb.int32))
def solve(L, R):
    A = np.zeros((R + 6, R + 6), dtype=np.float64)
    v = np.zeros(R + 6, dtype=np.float64)
    for x in range(R + 6):
        A[x, x] = (R - L + 1) * (R - L + 1)
        v[x] = 5 * (R - L + 1) * (R - L + 1)

        Lp = max(x - 5, L)
        v[x] += (R - L + 1) * (Lp - L) * (L + Lp - 1) / 2.

        for a in range(Lp, R + 1):
            for b in range(L + x, R + x + 1):
                if a <= b:
                    v[x] += a
                    A[x, b - a] -= 1.
                else:
                    v[x] += b
                    A[x, a - b] -= 1.
    f = np.linalg.solve(A, v)
    return f[0] - 5

import time
ans = 0
K = 100
t0 = time.time()
# for i in range(10):
#     solve(1, 100)
# print(time.time() - t0)
for L in range(1, K):
    print(L, time.time() - t0)
    for R in range(L, K):
        ans += solve(L, R + 1)
print("%.6f" % ans)
